### ISSUE TITLE
Jar Signing doesn't check the integrity of the actual files.

### ISSUE THREAD

**Issue by Darkhax** [MinecraftForge/MinecraftForge#7500]
**Minecraft Version:** 1.16.4

**Forge Version:** 35.0.18

**Logs:** https://gist.github.com/Darkhax/f89490bdfb61cb50ea88d2155e720904

**Steps to Reproduce:**
 1. Download a mod that has been signed. 
 2. Open the JAR with something that will let you tamper with the files. Like 7Zip.
 3. Edit the pack.mcmeta or mods.toml file to have some random junk data.
 4. The mod will still load fine, display the original signature, and present no tampering in the logs.

**Description of issue:** Tampering with the files of a mod will not trigger the security exception or invalidate the displayed signature. 

[Example Mod Tampering](https://github.com/MinecraftForge/MinecraftForge/files/5559987/bookshelf-files.zip)

**Comment by sciwhiz12** [comment]
If my reading of the code is correct, the functionality included in FML/Forge currently only displays the fingerprint of the certificate used to sign the JAR. It does not currently do any verification on whether the JARs are tampered with.
The firing of exceptions over invalidated/untrusted signatures is not implemented at this time, and probably will be in a future update (once cpw gets more time, I suppose).
For now, this can only be used to verify a JAR was signed by a specific certificate (and the Distinguised Name of that certificate, through the Mod List screen), and nothing more.

**Comment by Darkhax** [comment]
@sciwhiz12 Unfortunately the current system can not even be used for those use cases. It only cares that the signature is self consistent, so you can just rip the signatures out of almost any jar and make any new jar appear to be signed by that signature. For example this is a version of Bookshelf that was "signed by the real MinecraftForge". 

![image](https://user-images.githubusercontent.com/2250798/99545561-6a092f80-29ad-11eb-864b-49b53d82b709.png)

**Comment by cpw** [comment]
Interesting

**Comment by cpw** [comment]
Reopening, because it's a useful reminder to finish off this..

**Comment by PugzAreCute** [comment]
Its 2021 and still no update? Come on, and this seems like a major security flaw to me

**Comment by DaemonUmbra** [comment]
I'm not sure what you expect this feature to do that you can't do simply by being careful about where you get your mods.

**Comment by LexManos** [comment]
@cpw may address this eventually, however jar signing has never been a valid security feature. Be it in base java or the hacky classloader things we have to do. There is NO guarantee of security when running arbitrary code you download from the internet. This is purely informative.

**Comment by Darkhax** [comment]
Since it's been a while I wanted to post a follow up to this issue and respond to some of the comments about it.

<https://twitter.com/LexManos/status/1330217202025455620>
> To clarify, Forge is secure. As best as a random jar that we don't control the full chain, and explicitly load unsigned coremods, can be.
He's referring to people not liking the signatures being printed on the mod list.

The issue is caused by code signers being read from the JAR and then never being validated against the JAR contents. I personally like that the signature is shown on the mod list. Core mods are not required to reproduce this issue.

<https://github.com/MinecraftForge/MinecraftForge/issues/7500#issuecomment-729729248>
> It only cares that the signature is self consistent, so you can just rip the signatures out of almost any jar and make any new jar appear to be signed by that signature.

When I say "rip the signatures out" I mean the signature files in META-INF. You do not need to use reflection or other unsanctioned methods to change the signature displayed, it will simply accept any self consistent file.

<https://github.com/MinecraftForge/MinecraftForge/issues/7500#issuecomment-842828375>
> I'm not sure what you expect this feature to do that you can't do simply by being careful about where you get your mods.

There are two main issues here.

1. It reports back a false negative when JARs have been tampered with. This gives players a false sense of security and can lead to mod devs ruling out tampering when reviewing issue report logs.
2. It allows for easy impersonation of any dev who signs their JARs. As shown above it is trivial to impersonate Forge by simply copy pasting two files. 

<https://github.com/MinecraftForge/MinecraftForge/issues/7500#issuecomment-842862889>
> @cpw may address this eventually, however jar signing has never been a valid security feature. Be it in base java or the hacky classloader things we have to do. There is NO guarantee of security when running arbitrary code you download from the internet. This is purely informative.

You're right that full security is impossible when running arbitrary code from the internet. I just think the information displayed should have the bare minimum level of verification. Right now misconfigured builds, corrupted files, and not even script kiddie levels of tampering are capable of producing false negatives.

----
For now I have implemented my own solution using some [admittedly janky code](https://github.com/Darkhax-Minecraft/Minecraft-Modding-Template/blob/forge-1.16.5/src/main/java/net/darkhax/examplemod/ExampleMod.java#L39-L86). This is not a silver bullet by any means but does a good enough job to satisfy my concerns with Forge's implementation. 

I've already found one of my mods that was being incorrectly signed by myself due to misconfiguration, and another from the Chinese forums which changed the author, credits, license, and a few assets. Forge reported valid signatures for both.

**Comment by sciwhiz12** [comment]
> `@cpw` may address this eventually, however jar signing has never been a valid security feature. Be it in base java or the hacky classloader things we have to do. There is NO guarantee of security when running arbitrary code you download from the internet. This is purely informative.

Even if the information is considered purely informative, then what use is showing security-related information to the user if it can be easily forged? 

Even if, by your words, JAR signing is not a valid security feature to prevent potentially-malicious code from runnning on users' machines, having this information shown (both from the mods screen and the logs) can be misleading not only to users, but to developers who want to ensure that the JAR running on the user's machine is actually one which they have signed themselves and not modified nor manually built from unknown sources. _(Note that developers can see these signatures both from within their code and via logs that users provide)_

This might not be of critical importance, but the fact remains that a security-related feature is in Forge is showing user-facing information without properly verifying the information. If we are told that this feature is not expected to be actually used, then why did a non-functioning purely informative feature get pushed into production, and not kept in a branch for further development and testing?

**Comment by covers1624** [comment]
> Even if, by your words, JAR signing is not a valid security feature to prevent potentially-malicious code from runnning on users' machines, having this information shown (both from the mods screen and the logs) can be misleading not only to users, but to developers who want to ensure that the JAR running on the user's machine is actually one which they have signed themselves and not modified nor manually built from unknown sources. (Note that developers can see these signatures both from within their code and via logs that users provide)

My argument to this is mod authors should know their key signature and be able to identify when their mod is signed by a different key. It is feasibly impossible for someone to sign their jar with your key, without you giving them that key (Welcome to crypto).

On that same note, there is absolutely no chain of trust enforced by Forge, any key signing is self-signed, and as pointed out above, can be easily stripped from jars. Without some form of established Trust chain, the entire argument for 'Signing gives us security' is moot. This topic has come up before, and from what I remember, its not something that's opposed, its just more work to implement/maintain/manage. As people need to be able to submit their key to be signed, have illegal keys revoked, etc. 

Then there's still the argument of 'How do you make sure a key can only sign specific mods', you feasibly can't, as any hardcoded 'i require this signature' metadata stored within the mod can always be modified. Forge would then have to ship some form of 'These mod ids are owned by these keys' within their root public key, etc.

**Comment by sciwhiz12** [comment]
> My argument to this is mod authors should know their key signature and be able to identify when their mod is signed by a different key [...]

It is not the ability of the developer to determine if their key fingerprint is the same as reported that is in question (as any person capable of reading can read two lines of text and compare them letter by letter), but the ability of the system to report that the signatures on the JAR as being intact and consistent with the key and contents. 

Even if I were to recognize that the fingerprint of the key is the same as mine, it is pointless if I don't know if the signatures have been verified to be consistent with my key or the contents of the JAR (either via the information presented on screen or in logs, or via an exception thrown by the system upon reading inconsistent signatures).

> Without some form of established Trust chain, the entire argument for 'Signing gives us security' is moot.

First, a chain of trust already exists in the form of current CAs. A security-conscious developer could have a code-signing certificate from a known CA, which means the chain of trust can be validated by walking the certificate chain upto the root CA, and checking if the certificate store of the machine trusts it. Furthermore, a developer can include the chain of certificates when signing, which is usually done to include intermediate certificates (such is done in TLS).

Second, a chain of trust is not necessarily needed for security by signing. What is needed only are: the fingerprint of the signing certificate, the signatures are validy signed according to the certificate and public key, and that the digests from the signature match the contents of the JAR.

Given the fingerprint of the signing certificate, we can verify that the certificate used to sign belongs to the correct developer (since creating a certificate which has the same fingerprint is extremely difficult<sup>+</sup>). Once we verify the signing certificate belongs to the correct developer, we can then confirm that the signature is correct if the system says the signatures are validly signed by that certificate. Then, we can confirm that the JAR is untampered with if the signed digests are equal to the digests of the JAR's contents.

The chain of trust for certificates is necessary only for a client to ensure a certificate is valid (non-expired) and trusted (signed by a trusted CA). In this case, we don't need to verify if it is trusted by a CA, as the developer is the one who is checking that the certificate is correct via the fingerprint (and the developer trusts themselves to have the correct ceritficate).

> Then there's still the argument of 'How do you make sure a key can only sign specific mods'

You don't need to make sure a key can only sign specific mods. A developer should a) only sign their own mods, and b) be able to recognize their own certificate/key fingerprint. It is not necessarily the duty of Forge to ensure that those mods have signatures signed by specific authorities, its duty to _verify_ that the signatures in the JAR are actually _consistent_ with the data and the certificate given (which it currently fails at).

Being able to verify that certain mods are signed with certain keys is not what's being discussed here; what's being discussed is Forge not even verifying that the signatures packed with the JAR actually are consistent with both the contents of the JAR and the packed signing certificate. The argument of "what signs what" is solved via PKI, which is something which Forge doesn't need to solve on its own, as a PKI solution already exists: the Internet's PKI.

<sup>**+** Making a certificate with the same fingerprint entails having to find the exact amount of changes which _exactly_ transform the signature from another key into the same exact signature as the original, which would mean they've broken the [collision resistance](https://en.wikipedia.org/wiki/Collision_resistance) property of the cryptographic hash function in use, which is usually SHA-256. </sup>

**Comment by LexManos** [comment]
Alright so that's a wall of text with no real content, but i'll reply to a few of them before I go to bed {got to love early morning meetings}.
> You're right that full security is impossible when running arbitrary code from the internet. I just think the information displayed should have the bare minimum level of verification.

It was my understanding, and I assume cpw's as well that the JVM WOULD verify this while it was read from disc which is why he spent so much damn time trying to hack it into the JVM's normal signature loading system. Apparently it has been brought to my attention that no, the jvm doesn't verify when data is read, only when you do it through rather specific systems.

> For now I have implemented my own solution using some [admittedly janky code](https://github.com/Darkhax-Minecraft/Minecraft-Modding-Template/blob/forge-1.16.5/src/main/java/net/darkhax/examplemod/ExampleMod.java#L39-L86). This is not a silver bullet by any means but does a good enough job to satisfy my concerns with Forge's implementation.

Honestly, thats one of the only solution that could potentially be valid. Let the JVM flip its shit. The other option would be to read the hashes and compute them ourselves. Which avoids the exceptions/JRE internals which is nice. The only extra that needs to be added is to verify that the file you're reading is indeed signed. And if not add a 'This mod also contains unsigned files' because signatures are explicit lists, not garneted to be the entire contents of the jar. Tho this has the issue that it requires reading of ALL files in the jar. Which we do in our scanner code, so that would probably be the place to inject this verification.

> Without some form of established Trust chain, the entire argument for 'Signing gives us security' is moot. 

This is correct, and due to community drama, manpower, and the overall uselessness of the project this is not something we are going to do at all, ever.

> but the ability of the system to report that the signatures on the JAR as being intact and consistent with the key and contents.

Again, in theory it is, as the JVM *should* be running the same verifications it does during normal jar loading shit as any other program. It appears it is not and that is something we could look into.

> First, a chain of trust already exists in the form of current CAs. A security-conscious developer could have a code-signing certificate from a known CA, which means the chain of trust can be validated by walking the certificate chain upto the root CA,

This actually is not correct as certs are free and require little to no validation and are typically not meant for application development so anything that is signed via a trusted chain could be anybody. Again THIS HAS NOTHING WHAT SO EVER TO DO WITH SECURITY. If anyone thinks it provides any security in any way shape or form, then they are incorrect and not worth talking to as they do not understand the fundamentals of how Forge/The JVM/Signatures work. I am not going to continue this argument ANY information provided is purely informational and is not in any way related to the security of the system.

> its duty to verify that the signatures in the JAR are actually consistent with the data and the certificate given (which it currently fails at).

This is NOT Forge's DUTY, it is a potential informational feature. But definitely not a DUTY. As this information can be validated via the jar file itself and built in java tools. Anyone who actually gives a crap, or would know how to actually verify this information would know how to run the checks themselves. Again, the information displayed in the mod info pane is PURELY informational.

Anyways, this was a afternoon whim project for cpw, apparently his assumptions on how the system worked were wrong. He will get around to it eventually. Combined with the fact that it's not something we truly care about being 1) purely a informational text string in the gui and 2) Not even being functional on the default JRE that the Mojang launcher ships, its a stupid low priority.

It has ZERO effect on any security, it has ZERO effect for the average user. Anyone who DOES care can verify the information themselves. IF you do care so much rather then complain, properly articulate what your issues are and see if you can find a solution.

Honestly my personal opinion is the entire system should be ripped out because people see signatures and think it has something to do with security as evidenced but this entire conversation. But a *better* solution would be to slightly amend the informational line to say:
Signature: {} Manifest Only - For when the manifest only list itself and is valid
Signature: {} All Contents Valid - For when the signature list EVERY file in the jar, and each file is validated
Signature: {} Manifest contains files missing from jar - For when the manifest contains a valid signature, but the listed hashes contain a entry missing from the jar {Taking someone else's signature and shoving it into your jar, but not their files}
Signature: {} Manifest valid, but jar contains unsigned files. - Somewhat the same as the one above, but all listed files in the manifest are validated, the jar contains purely extra.

TLDR: Yes, IF we display this data, it should be 'more correct'. The information displayed is 100% correct as is. As the manifest is indeed signed by the signature that we display. It's just that the transitive files are not verified. And that could be valid extension to this system.

Now if someone wants to do this, then go for it. I'll talk to cpw when he next has some free time. But trying to label this as a actual security issue is not valid and will not be treated as one.

**Comment by cpw** [comment]
> Its 2021 and still no update? Come on, and this seems like a major security flaw to me

Yup. It turns out being busy with real life is FAR more important than this. If you don't like it, too bad

**Comment by cpw** [comment]
To those of you dramatis personae who believe this to be some kind of critical security flaw, don't be so ridiculous. Without a chain of trust it's just a statement about probable authorship. It's useful for detecting trivial forgeries and not much more.

**Comment by cpw** [comment]
Because people are spamming trivia and effluvia here, I am locking it, because _I knew this was a problem 6 months ago_ and it'll be fixed for 1.17. **HOPEFULLY**

**Comment by cpw** [comment]
This will be fixed in the java16 1.17 version of forge. There will be some significant enhancements around the behaviour here as well.. expect red warning triangles..
